internal/controller/rbac/provider/roles/reconciler.go

```diff
@@ -44,16 +45,36 @@ const (
 	timeout        = 2 * time.Minute
 	maxConcurrency = 5
 
-	errGetPR     = "cannot get ProviderRevision"
-	errListCRDs  = "cannot list CustomResourceDefinitions"
-	errApplyRole = "cannot apply ClusterRole"
+	errGetPR               = "cannot get ProviderRevision"
+	errListCRDs            = "cannot list CustomResourceDefinitions"
+	errApplyRole           = "cannot apply ClusterRole"
+	errValidatePermissions = "cannot validate permission requests"
+	errRejectedPermission  = "refusing to apply any RBAC roles due to request for disallowed permission"
 )
 
 // Event reasons.
 const (
 	reasonApplyRoles event.Reason = "ApplyClusterRoles"
 )
 
+// A PermissionRequestsValidator validates requested RBAC rules.
+type PermissionRequestsValidator interface {
+	// ValidatePermissionRequests validates the supplied slice of RBAC rules. It
+	// returns a slice of any rejected (i.e. disallowed) rules. It returns an
+	// error if it is unable to validate permission requests.
+	ValidatePermissionRequests(ctx context.Context, requested ...rbacv1.PolicyRule) ([]Rule, error)
+}
```

Comment:
It was a generic golang design question, as an aside, definitely not feedback that should impact whether this gets merged. Why is `PermissionRequestsValidator` an interface, and not just a callback fn?

Reply:
Ah, got it. It's probably mostly just a result of what felt more idiomatic to me when we started doing this. In some cases these interfaces have more than one method, or have implementations that could be cleaner as a more complex type (i.e. a struct with some state) rather than an enclosure callback.